### ISSUE TITLE
Limited options to use Esper features

### ISSUE THREAD

**Issue by gboege** [telefonicaid/perseo-fe#348]
The Esper documentation offers a very rich set of features: http://esper.espertech.com/release-8.2.0/reference-esper/html/index.html

Especially Context, but also NamedWindows and Tables are some of the powerful constructs in the Esper EPL:  http://esper.espertech.com/release-8.2.0/reference-esper/html/context.html

Currently, each rule that is submitted to the Perseo-fe is automatically put into the context of Service and Service-path. And due to this reason it is not possible to create other contexts manually. 

And it is not possible to "create" the other constructs because the EPL does not allow this keyword within/after the mandatory Context of Service and Service-path.

Yes, it is possible to put rules directly into perseo-core, but these rules are then unknown in perseo-fe and will be deleted in the setup interval. So this is not a valid work-around.

It would be great to learn about future strategic plans to open perseo-fe/-core to use the full feature set of the Esper CEP.

It could be like putting rules directly into perseo-core with a flag to not be deleted from perseo-fe. I assume it would be low effort compared to the great outcome.

**Comment by mrutid** [comment]
As far I know the fixed context of "Service/Service Path" is intended to provide some kind of separate scopes (separate rulesets and event streams) for different clients sharing a single Perseo CEP (this is a real scenario). I'm not sure of the implications of your proposal regarding this... maybe we could employ some kind of subcontext that allows us flexibility along with separated scopes "Service/ServicePath/CustomContext"... does this makes sense to you? Could you provide a PR/PoC/examples for this feature?

**Comment by gboege** [comment]
The idea of fixed context of "Service/Service Path" is valuable to separate the scopes of tenants and Service Paths. It makes it easy to start with the Esper engine behind perseo-core and perseo-fe. That's great. Thank you for this work and contribution. 

Currently there is a context AUTO-created for each "Service/Service Path" and each rule MUST start only with the select statement. Again, this is valuable and easy, but limited. 

Here the EPL syntax is mentioned: http://esper.espertech.com/release-8.2.0/reference-esper/html_single/index.html#epl-select-syntax

For each rule there will be the "Service/Service Path" auto-created and prepended to EACH rule to be injected via perseo-fe.

That means that it is impossible to inject a "create" EPL command. All these EPL objects can not be created: http://esper.espertech.com/release-8.2.0/reference-esper/html_single/index.html#gettingstarted_introepl
This is because these statements MUST start with a "create" statement, but this is not possible, because all rules will be auto-prepended with the auto-created context.

So, there are no manual schemas, variables, windows, tables, contexts, expressions (with UDF and i.e. Javascript functions) or indexes on tables.

A solution could be: 
a) To flag an incoming rule to be not prepended with the auto-context
b) To flag rules directly into perseo-core as not to be deleted in interval clean-up from perseo-fe.

So my issue is no new functionality but to unleash the ready-available power of the underlying EsperTech CEP engine for real advanced and powerful use cases. This could open up fantastic advanced FIWARE use cases that are currently limited

**Comment by fgalan** [comment]
@gboege maybe you could propose a PR to show how it would be, as @mrutid suggest? Don't bother with all the details a regular PR needs (documentation, tests, etc.), by the moment it would be just a kind of "proof of concept" so other developers can evaluate and provide feedback.

CC: @cblanco @aarranz (who also develop Perseo and maybe wants to provide feedback on this issue)

**Comment by gboege** [comment]
With some evening time, I was again checking the Perseo-Code for this issue. But not only the FE Code but also the Core-Code. And there I found the relevant code to delete the rules that are directly into the Core Engine via the Core API. 

And I found a already present solution to my issue. There is a config-parameter for the max-age of a rule that is not included in a refresh-all call. Haven't seen this before. 

With just setting this parameter in the docker-compose ENV to a very high value the deletion will not happen. My problem solved, all core features, also outside of the FE instructions usable. 

Issue closed.
